presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveNullRowInAggregation.java

```diff
@@ -0,0 +1,118 @@
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.facebook.presto.SystemSessionProperties.isRemoveNullRowInAggregation;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.plan.AggregationNode.Aggregation;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.AND;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IS_NULL;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static java.util.Objects.requireNonNull;
+
+public class RemoveNullRowInAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation();
+
+    private final StandardFunctionResolution functionResolution;
+    private final FunctionAndTypeManager functionManager;
+
+    public RemoveNullRowInAggregation(FunctionAndTypeManager functionAndTypeManager)
+    {
+        requireNonNull(functionAndTypeManager, "functionManager is null");
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        this.functionManager = functionAndTypeManager;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRemoveNullRowInAggregation(session);
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        PlanNode source = aggregationNode.getSource();
+        if (context.getLookup().resolveGroup(source).findFirst().filter(node -> node instanceof FilterNode)
+                .isPresent()) {
+            return Result.empty();
+        }
+        Collection<Aggregation> aggregations = aggregationNode.getAggregations().values();
+
+        //extract all the aggregation arguments
+        Set<RowExpression> rowExpressionSet = new HashSet<>();
+        for (Aggregation aggregation : aggregations) {
+            if (functionManager.getFunctionMetadata(aggregation.getFunctionHandle()).isCalledOnNullInput()) {
+                return Result.empty();
+            }
+            rowExpressionSet.addAll(aggregation.getCall().getArguments());
+        }
+
+        if (rowExpressionSet.isEmpty()) {
+            return Result.empty();
+        }
+
+        //Form all the ISNULL form
+        List<RowExpression> nullFormList = new ArrayList<>();
+        for (RowExpression rowExpression : rowExpressionSet) {
+            RowExpression nullForm = new SpecialFormExpression(IS_NULL, BOOLEAN, rowExpression);
+            nullFormList.add(nullForm);
+        }
+
+        RowExpression andForm = new SpecialFormExpression(AND, BOOLEAN, ImmutableList.copyOf(nullFormList));
```

Comment:
nit: I would use `ImmutableList.builder()` for `nullFormList`, call `add` in the loop, and then pass `nullFormList.build()` when generating the `andForm` expression. It prevents copying the list.

Reply:
fixed